provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/location/LocationController.java

```diff
@@ -1,3 +1,455 @@
+    List<CareSitesResponse.Value> careSiteResponsePages;
+    LocationWrapper currentPage;
+    totalRecords = 1;
+    int fromIndex;
+    int toIndex;
+    if (parameters.get("identifier") != null) {
+      String identifier = parameters.get("identifier").toArray()[0].toString();
+      locationWrapper.providerServicesResponse(ppmsClient.providerServicesById(identifier));
+      locationWrapper.providerResponse(
+          ProviderResponse.builder()
+              .value(
+                  Collections.singletonList(
+                      ProviderResponse.Value.builder()
+                          .providerIdentifier(
+                              Integer.parseInt(
+                                  parameters.get("identifier").toArray()[0].toString()))
+                          .build()))
+              .build());
+      if (locationWrapper.build().providerServicesResponse().value() == null
+          || locationWrapper.build().providerServicesResponse().value().isEmpty()) {
+        locationWrapper.providerResponse(ppmsClient.providersForId(identifier));
+        locationWrapper.careSitesResponse(ppmsClient.careSitesById(identifier));
+      }
+    } else if (parameters.get("name") != null) {
+      String name = parameters.get("name").toArray()[0].toString();
+      locationWrapper.providerResponse(
+          ppmsClient.providersForName(name.split("'")[0].split("/")[0].split("#")[0]));
+      totalRecords = locationWrapper.build().providerResponse().value().size();
+      fromIndex =
+          Math.min((page - 1) * count, locationWrapper.build().providerResponse().value().size());
+      toIndex =
+          Math.min((fromIndex + count), locationWrapper.build().providerResponse().value().size());
+
+      providerServicesResponsePages =
+          IntStream.range(fromIndex, toIndex)
+              .parallel()
+              .mapToObj(
+                  i -> {
+                    try {
+                      return ppmsClient.providerServicesByName(
+                          locationWrapper
+                              .build()
+                              .providerResponse()
+                              .value()
+                              .get(i)
+                              .name()
+                              .split("'")[0]
+                              .split("/")[0]
+                              .split("#")[0]);
+                    } catch (Exception e) {
+                      return ProviderServicesResponse.builder().build();
+                    }
+                  })
+              .collect(Collectors.toList());
+      providerResponsePages =
+          locationWrapper.build().providerResponse().value().subList(fromIndex, toIndex);
+      ProviderServicesResponse currentProviderServiceResponse;
+      CareSitesResponse currentCareSiteResponse;
+      for (int i = 0; i < providerServicesResponsePages.size(); i++) {
+        currentProviderServiceResponse = providerServicesResponsePages.get(i);
```

Comment:
Since we are making more calls to PPMS inside this loop, there is opportunity for more parallelization. The body of this loop should roughly become its own method so that it can cleanly be invoked inside a `parallelStream`, similar to above.